src/scene/forward-renderer.js

```diff
@@ -761,30 +916,28 @@ pc.extend(pc, function () {
                     var passes = 1;
                     var pass;
 
+                    shadowCam._node.setPosition(light._node.getPosition());
+                    shadowCam._node.setRotation(light._node.getRotation());
+                    // Camera's look down negative Z, and directional lights point down negative Y
+                    shadowCam._node.rotateLocal(-90, 0, 0);
+
                     if (type === pc.LIGHTTYPE_DIRECTIONAL) {
-                        var shadowDistance = light.getShadowDistance();
-
-                        // 1. Starting at the centroid of the view frustum, back up in the opposite
-                        // direction of the light by a certain amount. This will be our temporary
-                        // working position.
-                        _getFrustumCentroid(scene, camera, shadowDistance, this.centroid);
-                        shadowCam._node.setPosition(this.centroid);
-                        shadowCam._node.setRotation(light._node.getRotation());
-                        // Camera's look down negative Z, and directional lights point down negative Y
-                        shadowCam._node.rotateLocal(-90, 0, 0);
-
-                        // 2. Transform the 8 corners of the camera frustum into the shadow camera's
-                        // view space
-                        _getFrustumPoints(scene, camera, shadowDistance, frustumPoints);
-                        shadowCamWtm.copy(shadowCam._node.getWorldTransform());
-                        var worldToShadowCam = shadowCamWtm.invert();
-                        var camToWorld = camera._node.worldTransform;
-                        c2sc.mul2(worldToShadowCam, camToWorld);
+                        // 1. Get the frustum of the camera
+                        _getFrustumPoints(camera, light.getShadowDistance()||camera.getFarClip(), frustumPoints);
+
+                        // 2. Firgure it out the maximum diagonal of the frustum in light's projected space.
+                        var diagonal = new pc.Vec3();

```

Comment:
Can you remove "new" from the main loop? It's better to do all memory allocations in advance.